library/core/src/any.rs

```diff
@@ -700,3 +772,302 @@ pub const fn type_name<T: ?Sized>() -> &'static str {
+    /// # Examples
+    ///
+    /// Provides a reference to a field with type `String` as a `&str`, and a value of
+    /// type `i32`.
+    ///
+    /// ```rust
+    /// # #![feature(provide_any)]
+    /// use std::any::{Provider, Demand};
+    /// # struct SomeConcreteType { field: String, num_field: i32 }
+    ///
+    /// impl Provider for SomeConcreteType {
+    ///     fn provide<'a>(&'a self, demand: &mut Demand<'a>) {
+    ///         demand.provide_ref::<str>(&self.field)
+    ///             .provide_value::<i32, _>(|| self.num_field);
+    ///     }
+    /// }
+    /// ```
+    #[unstable(feature = "provide_any", issue = "96024")]
+    fn provide<'a>(&'a self, demand: &mut Demand<'a>);
+}
+
+/// Request a value from the `Provider`.
+///
+/// # Examples
+///
+/// Get a string value from a provider.
+///
+/// ```rust
+/// # #![feature(provide_any)]
+/// use std::any::{Provider, request_value};
+///
+/// fn get_string<P: Provider>(provider: &P) -> String {
+///     request_value::<String, _>(provider).unwrap()
+/// }
+/// ```
+#[unstable(feature = "provide_any", issue = "96024")]
+pub fn request_value<'a, T, P>(provider: &'a P) -> Option<T>
+where
+    T: 'static,
+    P: Provider + ?Sized,
+{
+    request_by_type_tag::<'a, tags::Value<T>, P>(provider)
+}
+
+/// Request a reference from the `Provider`.
+///
+/// # Examples
+///
+/// Get a string reference from a provider.
+///
+/// ```rust
+/// # #![feature(provide_any)]
+/// use std::any::{Provider, request_ref};
+///
+/// fn get_str<P: Provider>(provider: &P) -> &str {
+///     request_ref::<str, _>(provider).unwrap()
+/// }
+/// ```
+#[unstable(feature = "provide_any", issue = "96024")]
+pub fn request_ref<'a, T, P>(provider: &'a P) -> Option<&'a T>
+where
+    T: 'static + ?Sized,
+    P: Provider + ?Sized,
+{
+    request_by_type_tag::<'a, tags::Ref<tags::MaybeSizedValue<T>>, P>(provider)
+}
+
+/// Request a specific value by tag from the `Provider`.
+fn request_by_type_tag<'a, I, P>(provider: &'a P) -> Option<I::Reified>
+where
+    I: tags::Type<'a>,
+    P: Provider + ?Sized,
+{
+    let mut tagged = TaggedOption::<'a, I>(None);
+    provider.provide(tagged.as_demand());
+    tagged.0
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// Demand and its methods
+///////////////////////////////////////////////////////////////////////////////
+
+/// A helper object for providing data by type.
+///
+/// A data provider provides values by calling this type's provide methods.
+#[unstable(feature = "provide_any", issue = "96024")]
+#[repr(transparent)]
+pub struct Demand<'a>(dyn Erased<'a> + 'a);
+
+impl<'a> Demand<'a> {
+    /// Create a new `&mut Demand` from a `&mut dyn Erased` trait object.
+    fn new<'b>(erased: &'b mut (dyn Erased<'a> + 'a)) -> &'b mut Demand<'a> {
+        // SAFETY: transmuting `&mut (dyn Erased<'a> + 'a)` to `&mut Demand<'a>` is safe since
+        // `Demand` is repr(transparent).
+        unsafe { transmute(erased) }
```

Comment:
This `transmute` can be replaced by pointer casting:
```suggestion
        unsafe { &mut *(erased as *mut dyn Erased as *mut Demand) }
```